### PR TITLE
Updates .use and .useError to a new pattern

### PR DESCRIPTION
**Pull request by vmike245**
.use has moved to .then
.useError has moved to .catch

They now both use the same stack, so it is possible to recover from an error state and continue execution.

### CODE REVIEW

File: lib/Route.js [hunk 1]
```diff
@@ -181,7 +195,14 @@ class Route {
     }
     try {
       request.routeParams = this._parsePath(request.path);
-      return Q(this.handler(options));
+      if (this.errorHandler) {
+        return Q(Q(this.handler(options)).catch((error) => {
```
Comment (autoric) [inline lib/Route.js]: Too many Q's?

File: lib/Route.js [hunk 2]
```diff
@@ -181,7 +195,14 @@ class Route {
...
+          return this.errorHandler({ request, response, error });
+        }));
+      }
+      else {
+        return Q(this.handler(options));
```
Comment (autoric) [inline lib/Route.js]: This whole block (lines 198-205) could be cleaned up into a single promise chain.
```
let handlerPromise = Q(this.handler(options));
if (this.errorHandler) {
  handlerPromise = handlerPromise.catch((error) => {
    return this.errorHandler({ request, response, error });
  });
}
return handlerPromise;
```

File: lib/App.js [hunk 3]
```diff
@@ -43,7 +53,7 @@ class App {
    *  used as the handler. If a plain object is provided, it will be run through the Route constructor.
    * @returns {App} - app instance for chaining
    */
-  use(route) {
+  then(route) {
```
Comment (autoric) [inline lib/App.js]: then and catch methods look pmuch exactly the same at this point. Lets delegate to implementation method of `_addMiddleware`

File: lib/App.js [hunk 4]
```diff
@@ -51,7 +61,7 @@ class App {
       route = new Route(route);
     }
 
-    this._middleware.push(route);
+    this._middleware.push({ route, errorMiddleware : false });
```
Comment (autoric) [inline lib/App.js]: Rather than pushing a bool of `errorMiddleware`, let's just push `{method : 'then' | 'catch' }`

File: lib/App.js [hunk 5]
```diff
@@ -109,22 +133,26 @@ class App {
...
-      promise = promise.then(() => {
-        if (!response.ended) {
-          return route.execute({ request, response });
-        }
-      });
-    });
-    _.each(this._errorMiddleware, (route) => {
-      if (!route.isMatch(request)) {
-        return;
+    _.each(this._middleware, (registeredRoute) => {
```
Comment (autoric) [inline lib/App.js]: Rename var `registeredRoute` to `middleware`. Cause you know...I'm iterating over middleware.

File: lib/App.js [hunk 6]
```diff
@@ -109,22 +133,26 @@ class App {
...
+      const route = registeredRoute.route;
+      const errorMiddleware = registeredRoute.errorMiddleware;
```
Comment (autoric) [inline lib/App.js]: This should now be a string of method equal to `then` or `catch` instead of a bool, SOOOO....

File: lib/App.js [hunk 7]
```diff
@@ -109,22 +133,26 @@ class App {
...
+      if (!errorMiddleware) {
```
Comment (autoric) [inline lib/App.js]: instead of switching on bool, just do:
```
promise = promise[method](() => {/*...*/})
```

File: test/App.test.js [hunk 8]
```diff
@@ -342,4 +342,106 @@ describe('App', () => {
...
+        expect(middleware2.execute).to.not.have.been.called;
+        expect(middleware3.execute).to.not.have.been.called;
+        expect(errMiddleware1.execute).to.have.been.calledOnce;
+        expect(errMiddleware2.execute).to.have.been.calledOnce;
+        expect(middleware4.execute).to.have.been.calledOnce;
+      });
+    });
+
+    it('should bubble up error if last error middleware throws error', () => {
+      middleware1.execute.returnsPromise();
```
Comment (autoric) [inline test/App.test.js]: move all of the `.returnsPromise` calls to the test setup.

File: test/Route.test.js [hunk 9]
```diff
@@ -410,6 +423,75 @@ describe('A Route', () => {
...
+        let errorHandler;
+
+        beforeEach(() => {
+          errorHandler = sinon.stub();
+          opts    = { path : '/', methods : 'GET', handler : handler, errorHandler : errorHandler };
+          route   = new Route(opts);
+        });
+
+        it('if the handler resolves then error handler should not be run', () => {
+          handler.returnsPromise();
```
Comment (autoric) [inline test/Route.test.js]: Again put the`returnsPromise` calls once at the setup.

File: lib/App.js (hunk 7 above)
Comment (vmike245) [inline lib/App.js]: Is this possible? The else block has a separate check where `if (!route.isMatch(request)) {
        return;
          }` and then error is being passed in if it is a catch. I moving the isMatch code in front of the then causes about 10 test failures. Is there something I am missing?

**Pull request merged**